README.md

```diff
@@ -78,9 +78,112 @@ If you're bootstrapping a Juju controller within a `multipass` VM you will not h
+- We limit the use of snapshots except for where the snapshot updates can be easily verified by a reviewer.
+- Add assertions for the explicit content you're expecting. This allows changes to things that may not be relevant to the test like classNames, attributes, etc.
+- When searching for an element, use element selectors where possible and add a data attribute when not. We follow the format of `data-test="..."` or `data-test-<specifier>="..."` ex) `<a data-test-column="priority">High</a>`.
+- Test labels should be specific and representative of the test content. Create additional tests if you need assertions that do not apply to the label.
+- Do not mock components unless absolutely necessary. The data dump mentioned below should contain sufficient data to render the full component tree in the test.
+
+#### Mocking data
+
+The file [complete-redux-store-dump.js](src/testing/complete-redux-store-dump.js) contains a sanitized redux dump from a Dashboard instance with many models in different states. When writing a test use the real data provided by this file when generating a redux store that will provide the data to your component or library. Because every test uses the same data set this allows us to see when a library or component will fail if the data that's saved to the redux store, or a selector is modified.
+
+> **If you modify the data that is saved into the redux store then you must update this file**
+
+If you want to test states that are not included in the data dump, or if you want to see how a component will act in different data states you can modify it locally within the test. If you must modify some value see if there is value in adding another entry into the data set for others to use.
+
+#### Test Snippits
+
+The typical imports and test case setup looks like the following. Note that we create a shared mockStore factory at the top of each test suite.
+
+```
+import React from "react";
+import { MemoryRouter } from "react-router";
+import { mount } from "enzyme";
+import { Provider } from "react-redux";
+import configureStore from "redux-mock-store";
+import dataDump from "../../testing/complete-redux-store-dump";
+
+const mockStore = configureStore([]);
+
+describe("ComponentName", () => {
+
+ it('', () => {
+   // test case
+ });
+
+ //... more tests
+
+});
+```
+
+Setting up a typical component test involves creating a mock store and passing that to the Provider so that any component you're testing can still use its built-in redux selectors.
+
+```
+const store = mockStore(dataDump);
+const wrapper = mount(
+  <MemoryRouter>
+    <Provider store={store}>
+      <StatusGroup />
+    </Provider>
+  </MemoryRouter>
+);
+```
+
+If you need to modify the data used in the application, clone the data dump in each test so that you do not run the risk of overriding data in subsequent tests. You should only update the values directly.
+
+```
+const clonedData = cloneDeep(dataDump);
```

Comment:
Just curious - does this add a significant amount of time to running tests?

Reply:
We haven't had any performance issues with our tests so tbh I didn't test it but I took this as an opportunity to run some micro-benchmarks and the averages are:

19ms to clone the dataDump which is 11,500 lines of heavily nested objects.
95ms to mount the `ModelTableList` component to the DOM in `jsdom`
18ms to run `wrapper.find` an element in the DOM.

So it looks like it doesn't really have any appreciable performance hit in the tests but we also discourage this type of modification where possible so it's only used 4 times throughout the whole test suite.